src/manager/runner_scaler.py

```diff
@@ -0,0 +1,215 @@
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class RunnerInfo:
+    """Information on the runners.
+
+    Attributes:
+        online: The number of runner in online state.
+        busy: The number of the runner in busy state.
+        offline: The number of runner in offline state.
+        unknown: The number of runner in unknown state.
+        runners: The names of the online runners.
+        busy_runners: The names of the busy runners.
+    """
+
+    online: int
+    busy: int
+    offline: int
+    unknown: int
+    runners: tuple[str, ...]
+    busy_runners: tuple[str, ...]
+
+
+class RunnerScaler:
+    """Manage the reconcile of runners."""
+
+    def __init__(self, runner_manager: RunnerManager, reactive_config: ReactiveConfig | None):
+        """Construct the object.
+
+        Args:
+            runner_manager: The RunnerManager to perform runner reconcile.
+            reactive_config: Reactive runner configuration.
+        """
+        self._manager = runner_manager
+        self._reactive_config = reactive_config
+
+    def get_runner_info(self) -> RunnerInfo:
+        """Get information on the runners.
+
+        Returns:
+            The information on the runners.
+        """
+        runner_list = self._manager.get_runners()
+        online = 0
+        busy = 0
+        offline = 0
+        unknown = 0
+        online_runners = []
+        busy_runners = []
+        for runner in runner_list:
+            match runner.github_state:
+                case GitHubRunnerState.BUSY:
+                    online += 1
+                    online_runners.append(runner.name)
+                    busy += 1
+                    busy_runners.append(runner.name)
+                case GitHubRunnerState.IDLE:
+                    online += 1
+                    online_runners.append(runner.name)
+                case GitHubRunnerState.OFFLINE:
+                    offline += 1
+                case _:
+                    unknown += 1
+        return RunnerInfo(
+            online=online,
+            busy=busy,
+            offline=offline,
+            unknown=unknown,
+            runners=tuple(online_runners),
+            busy_runners=tuple(busy_runners),
+        )
+
+    def flush(self, flush_mode: FlushMode = FlushMode.FLUSH_IDLE) -> int:
+        """Flush the runners.
+
+        Args:
+            flush_mode: Determines the types of runner to be flushed.
+
+        Returns:
+            Number of runners flushed.
+        """
+        metric_stats = self._manager.cleanup()
+        delete_metric_stats = self._manager.flush_runners(flush_mode=flush_mode)
+        events = set(delete_metric_stats.keys()) | set(metric_stats.keys())
+        metric_stats = {
+            event_name: delete_metric_stats.get(event_name, 0) + metric_stats.get(event_name, 0)
+            for event_name in events
+        }
+        return metric_stats.get(metric_events.RunnerStop, 0)
+
+    def reconcile(self, quantity: int) -> int:
```

Comment:
This method looks quite big, takes a long time to read, and will probably have to be broken down soon because of its cognitive complexity. Perhaps we are already breaking it up.

Reply:
Are you planning on refactoring as part of your PR, or should I do it here.

Reply:
Its fine for me to let it as it is, but we'll probably have to refactor it once we touch it next time.